source/solvers/fluid_dynamics_matrix_free_operators.cc

```diff
@@ -830,6 +986,201 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
+
+  //  std::cout << "Integrating over face " << face << " assemble residual "
+  //            << assemble_residual << " number of quadrature points "
+  //            << integrator.n_q_points << std::endl;
+
+  const double beta = this->boundary_conditions.beta[boundary_index];
+
+
+  const unsigned fe_degree =
+    matrix_free.get_dof_handler().get_fe().tensor_degree();
+
+  VectorizedArray<number> cell_size = 1.0;
+
+  for (unsigned int v = 0;
+       v < matrix_free.n_active_entries_per_face_batch(face);
+       ++v)
+    {
+      const auto [cell_it, _] = matrix_free.get_face_iterator(face, v);
+
+      if (dim == 2)
+        cell_size[v] = std::sqrt(4. * cell_it->measure() / M_PI) / fe_degree;
+      else if (dim == 3)
+        cell_size[v] =
+          std::pow(6 * cell_it->measure() / M_PI, 1. / 3.) / fe_degree;
+    }
+
+  VectorizedArray<number> penalty_parameter =
+    beta / std::pow(cell_size, static_cast<number>(fe_degree + 1));
+
+  integrator.evaluate(EvaluationFlags::EvaluationFlags::values |
+                      EvaluationFlags::EvaluationFlags::gradients);
+
+  for (const auto q : integrator.quadrature_point_indices())
+    {
+      typename FEFaceIntegrator::value_type    value_result    = {};
+      typename FEFaceIntegrator::gradient_type gradient_result = {};
+
+      const auto normal_vector = integrator.get_normal_vector(q);
+
+      const auto value    = integrator.get_value(q);
+      const auto gradient = integrator.get_gradient(q);
+
+
+      for (unsigned int d = 0; d < dim; ++d)
+        { // Assemble (v,beta (u)) for the main penalization
+          value_result[d] += penalty_parameter * value[d];
+
+
+          // Assemble ν(v,∇δu·n)
+          for (unsigned int i = 0; i < dim; ++i)
+            value_result[d] -=
+              kinematic_viscosity * gradient[d][i] * normal_vector[i];
+          //
+          // // // Assemble ν(∇v·n,δu)
+          for (unsigned int i = 0; i < dim; ++i)
+            gradient_result[d][i] -=
+              kinematic_viscosity * value[d] * normal_vector[i];
+        }
+
+
+      // If assemble_residual, we need to also assemble the target value of the
+      // weak dirichlet boundary condition which obtained from the mu_parser
+      // function contained in the boundary_conditions object
+      if constexpr (assemble_residual)
+        {
+          Point<dim, VectorizedArray<number>> point_batch =
+            integrator.quadrature_point(q);
+
+          // std::cout << " The points are " << point_batch << std::endl;
+
+          // std::cout << " Value is : " << value << std::endl;
+
+          Tensor<1, dim, VectorizedArray<number>> velocity_value;
+
+          velocity_value[0] = evaluate_function<dim, number>(
+            boundary_conditions.bcFunctions[boundary_index].u, point_batch);
+
+          velocity_value[1] = evaluate_function<dim, number>(
+            boundary_conditions.bcFunctions[boundary_index].v, point_batch);
```

Comment:
This can be also precomputed, since the information is used in multiple newton steps.

Reply:
Currently I lumped it with the stabilization evaluation because I did not want to create another indirection. Since this is pre-computed for only a few faces, I don't think this will make a big difference.